cEP-0002.md

```diff
@@ -0,0 +1,403 @@
+                                                         bear,
+                                                         dep_resolver,
+                                                         blocking_tasks,
+                                                         event_loop,
+                                                         executor))
+
+
+def finish_task(bear,
+                dep_resolver,
+                blocking_tasks,
+                event_loop,
+                executor,
+                task):
+    """
+    The callback for when a task of a bear completes. It is responsible for
+    checking if the bear completed it's execution and the handling of the
+    result generated by the task.
+    :param bear:           The bear that the task belongs to.
+    :param dep_resolver:   The object that keeps track of dependencies.
+    :param blocking_tasks: Dictionary that keeps track of the remaining tasks
+                           of each bear.
+    :param event_loop:     The asyncio event loop.
+    :param executor:       The process pool to which the bear tasks are
+                           scheduled.
+    :param task:           The task that completed.
+    """
+    log = logging.getLogger('Consumer task')
+    log.info(bear)
+    blocking_tasks[bear].remove(task)
+    if not blocking_tasks[bear]:
+        resolved_bears = dep_resolver.resolve(bear)
+        if resolved_bears:
+            schedule_bears(resolved_bears,
+                           dep_resolver,
+                           event_loop,
+                           blocking_tasks,
+                           executor)
+        del blocking_tasks[bear]
+    if not blocking_tasks:
+        event_loop.stop()
+
+
+def main():
+    # Configure logging to show the name of the thread where the log
+    # message originates.
+    logging.basicConfig(
+        level=logging.INFO,
+        format='%(process)5s %(name)18s: %(message)s',
+        stream=sys.stderr,
+    )
+
+    dep_resolver = Dependencies()
+
+    # Create a limited thread pool.
+    executor = concurrent.futures.ProcessPoolExecutor(
+        max_workers=get_cpu_count(),
+    )
+    event_loop = asyncio.get_event_loop()
+
+    file_list = collect_files(["~/Documents/gsoc/coala/*.py",
+                               "~/Documents/gsoc/coala/coalib/*.py",
+                               "~/Documents/gsoc/coala/coalib/**/*.py"], None)
+
+    fileproxies = [FileProxy(file) for file in file_list]
+
+    bears = [TestBear(fileproxies),
+             AnotherBear(fileproxies),
+             AnotherDependentBear(fileproxies),
+             DependentBear(fileproxies)]
+
+    blocking_tasks = {}
+    dep_resolver.add_bear_dependencies(bears)
+
+    schedule_bears(bears,
+                   dep_resolver,
+                   event_loop,
+                   blocking_tasks,
+                   executor)
+    try:
+        event_loop.run_forever()
+    finally:
+        event_loop.close()
```

Comment:
The bears will throw exceptions in a different process. Will those try catch link back to this?